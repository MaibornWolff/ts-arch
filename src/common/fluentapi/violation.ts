export type Violation = {
	message: string;
	details: any;
}
