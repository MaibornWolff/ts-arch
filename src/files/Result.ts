import { Noun } from "./noun/Noun"

export interface ResultEntry {
	subject: Noun
	object?: Noun
	info: string
	pass: boolean
}

export class Result {
	private entries: Array<ResultEntry> = []
	private hasForcePassed = false
	private hasForceFailed = false

	public forcePass() {
		this.hasForcePassed = true
		this.hasForceFailed = false
	}

	public forceFail() {
		this.hasForcePassed = false
		this.hasForceFailed = true
	}

	public addEntry(entry: ResultEntry) {
		this.entries.push(entry)
	}

	public getEntries(): Array<ResultEntry> {
		return this.entries
	}

	public hasRulePassed(): boolean {
		if (this.hasForcePassed) {
			return true
		}
		if (this.hasForceFailed) {
			return false
		}
		let result = true
		this.entries.forEach(r => {
			if (!r.pass) {
				result = false
			}
		})
		return result
	}
}
