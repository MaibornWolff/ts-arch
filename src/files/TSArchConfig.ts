export interface TSArchConfig {
	ignore: IgnoreConfig;
}
export interface IgnoreConfig {
	declarations: boolean;
	nodeModules: boolean;
	js: boolean;
}
