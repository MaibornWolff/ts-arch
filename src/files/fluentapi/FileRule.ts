import {Result} from "neverthrow";
import {ViolatingFile} from "../assertions/matchingFiles";

export interface FileRule {
	check(): Promise<Result<ViolatingFile[], string>>
}
