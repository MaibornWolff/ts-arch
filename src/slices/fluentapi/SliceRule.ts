import {Result} from "neverthrow";
import {ViolatingEdge} from "../assertions/admissibleEdges";

export interface EdgeRule {
	check(): Promise<Result<ViolatingEdge[], string>>
}
