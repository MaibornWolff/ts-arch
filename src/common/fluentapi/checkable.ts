import {Violation} from "./violation";

export interface Checkable {
	check(): Promise<Violation[]>
}
