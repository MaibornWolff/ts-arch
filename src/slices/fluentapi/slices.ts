import { generateRule } from "../uml/generateRules"
import { sliceByPattern } from "../projections/slicing"
import { project } from "../processing/project"
import { extractGraph } from "../../common/extraction/extractGraph"
import {
	gatherPositiveViolations,
	ViolatingEdge,
	Rule,
	gatherViolations
} from "../assertions/admissibleEdges"
import { Result, err, ok } from "neverthrow"
import fs from "fs"
import {EdgeRule} from "./SliceRule";

export function slicesOfProject(filename?: string): SliceConditionBuilder {
	return new SliceConditionBuilder(filename)
}

export class SliceConditionBuilder {
	constructor(readonly filename?: string) {}

	public definedBy(pattern: string): SlicesShouldCondition {
		return new SlicesShouldCondition(this, pattern)
	}
}

export class SlicesShouldCondition {
	constructor(readonly sliceCondition: SliceConditionBuilder, readonly pattern: string) {}
	public should(): PositiveConditionBuilder {
		return new PositiveConditionBuilder(this)
	}

	public shouldNot(): NegativeSliceCondition {
		return new NegativeSliceCondition(this, [])
	}
}

export class NegativeSliceCondition implements EdgeRule {
	constructor(
		readonly slicesShouldcondition: SlicesShouldCondition,
		private readonly forbiddenEdges: Rule[]
	) {}

	public containDependency(from: string, to: string): NegativeSliceCondition {
		return new NegativeSliceCondition(this.slicesShouldcondition, [
			...this.forbiddenEdges,
			{ source: from, target: to }
		])
	}

	public async check(): Promise<Result<ViolatingEdge[], string>> {
		const graph = await extractGraph(this.slicesShouldcondition.sliceCondition.filename)
		if (graph.isErr()) {
			return err(graph.error)
		}
		const mapFunction = sliceByPattern(this.slicesShouldcondition.pattern)

		const mapped = project(graph.value, mapFunction)
		return ok(gatherViolations(mapped, this.forbiddenEdges))
	}
}

export class PositiveConditionBuilder {
	constructor(readonly slicesShouldCondition: SlicesShouldCondition) {}

	public adhereToDiagram(diagram: string): PositiveSliceCondition {
		return new PositiveSliceCondition(this, { diagram })
	}

	public adhereToDiagramInFile(filename: string): PositiveSliceCondition {
		return new PositiveSliceCondition(this, { filename })
	}
}

export class PositiveSliceCondition implements EdgeRule{
	constructor(
		readonly positiveConditionBuilder: PositiveConditionBuilder,
		readonly diagram: { filename?: string; diagram?: string }
	) {}

	public async check(): Promise<Result<ViolatingEdge[], string>> {
		const graph = await extractGraph(
			this.positiveConditionBuilder.slicesShouldCondition.sliceCondition.filename
		)
		let diagram = this.diagram.diagram
		const filename = this.diagram.filename
		if (graph.isErr()) {
			return err(graph.error)
		}

		if (diagram === undefined) {
			if (filename !== undefined) {
				diagram = fs.readFileSync(filename).toString()
			} else {
				return err("No diagram provided")
			}
		}
		const rules = generateRule(diagram)

		const mapFunction = sliceByPattern(this.positiveConditionBuilder.slicesShouldCondition.pattern)

		const mapped = project(graph.value, mapFunction)
		return ok(gatherPositiveViolations(mapped, rules))
	}
}
