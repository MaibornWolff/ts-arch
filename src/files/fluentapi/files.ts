import {extractGraph} from "../../common/extraction/extractGraph"
import {gatherRegexMatchingViolations} from "../assertions/matchingFiles";
import {projectToNodes} from "../processing/project";
import {RegexFactory} from "../domain/RegexFactory";
import {Checkable} from "../../common/fluentapi/checkable";
import {Violation} from "../../common/fluentapi/violation";

export function filesOfProject(tsConfigFilePath?: string): FileConditionBuilder {
	return new FileConditionBuilder(tsConfigFilePath)
}

export class FileConditionBuilder {
	constructor(readonly tsConfigFilePath?: string) {}

	public matchingPattern(pattern: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [pattern])
	}

	public withName(name: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [RegexFactory.fileNameMatcher(name)])
	}

	public inFolder(folder: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [RegexFactory.folderMatcher(folder)])
	}
}

export class FilesShouldCondition {
	constructor(readonly fileCondition: FileConditionBuilder, readonly patterns: string[]) {}

	public should(): MatchPatternFileConditionBuilder {
		return new MatchPatternFileConditionBuilder(this, false)
	}
	public shouldNot(): MatchPatternFileConditionBuilder {
		return new MatchPatternFileConditionBuilder(this, true)
	}

	public matchingPattern(pattern: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [...this.patterns, pattern])
	}

	public withName(name: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [...this.patterns, RegexFactory.fileNameMatcher(name)])
	}

	public inFolder(folder: string): FilesShouldCondition {
		return new FilesShouldCondition(this, [...this.patterns, RegexFactory.folderMatcher(folder)])
	}
}

export class MatchPatternFileConditionBuilder {
	constructor(readonly filesShouldCondition: FilesShouldCondition,
				readonly isNegated: boolean) {}

	public matchPattern(pattern: string): MatchPatternFileCondition {
		return new MatchPatternFileCondition(this, pattern)
	}

	public beInFolder(folder: string): MatchPatternFileCondition {
		return new MatchPatternFileCondition(this, RegexFactory.folderMatcher(folder))
	}
}

export class MatchPatternFileCondition implements Checkable {
	constructor(
		readonly matchPatternFileConditionBuilder: MatchPatternFileConditionBuilder,
		readonly pattern: string
	) {}

	public async check(): Promise<Violation[]> {
		const graph = await extractGraph(
			this.matchPatternFileConditionBuilder.filesShouldCondition.fileCondition.tsConfigFilePath
		)

		const projectedNodes = projectToNodes(graph)

		return gatherRegexMatchingViolations(projectedNodes,
			this.pattern,
			this.matchPatternFileConditionBuilder.filesShouldCondition.patterns,
			this.matchPatternFileConditionBuilder.isNegated)
	}
}
