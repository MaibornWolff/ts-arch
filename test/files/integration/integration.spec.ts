import {filesOfProject} from "../../../src/files/fluentapi/files";

describe("Integration test", () => {

	it("repetitive filtering of files is possible", async () => {
		const violations = await filesOfProject(__dirname + "/samples/filenamingsample/tsconfig.json")
			.matchingPattern(".*Controller\.ts")
			.matchingPattern(".*troller\.ts")
			.inFolder("services")
			.shouldNot()
			.matchPattern(".*\\\\services\\\\.*") // TODO we should unify path notations
			.check()

		expect(violations).toEqual([{"label": "src\\services\\SomeController.ts", "rule": "should not match regex '.*\\\\services\\\\.*'"}])
	})

	it("controllers should not be in services", async () => {
		const violations = await filesOfProject(__dirname + "/samples/filenamingsample/tsconfig.json")
			.matchingPattern(".*Controller\.ts")
			.shouldNot()
			.beInFolder("services") // TODO we should unify path notations
			.check()

		expect(violations).toMatchSnapshot()
	})

	it("services should be named Service", async () => {

		const violations = await filesOfProject(__dirname + "/samples/filenamingsample/tsconfig.json")
			.inFolder("services") // TODO we should unify path notations
			.should()
			.matchPattern(".*Service\.ts")
			.check()

		expect(violations).toMatchSnapshot()
	})
})
