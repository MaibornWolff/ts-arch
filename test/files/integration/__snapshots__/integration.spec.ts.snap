// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test controllers should not be in services 1`] = `
Array [
  Object {
    "details": Object {
      "file": Object {
        "label": "src\\\\services\\\\SomeController.ts",
      },
      "pattern": ".*\\\\\\\\services\\\\\\\\.*",
    },
    "message": "src\\\\services\\\\SomeController.ts should not match regex '.*\\\\\\\\services\\\\\\\\.*'",
  },
]
`;

exports[`Integration test services should be named Service 1`] = `
Array [
  Object {
    "details": Object {
      "file": Object {
        "label": "src\\\\services\\\\SomeController.ts",
      },
      "pattern": ".*Service.ts",
    },
    "message": "src\\\\services\\\\SomeController.ts should match regex '.*Service.ts'",
  },
]
`;
